### ISSUE TITLE
[TF 2.0] Allow declaring dynamic tf.Variable shapes

### ISSUE THREAD

**Issue by alanhdu** [tensorflow/tensorflow#26325]
In the current TensorFlow 2.0 beta, `tf.Variables` inherits its shape from the initial value that it's provided:

```python
>>> tf.Variable(tf.zeros([16, 32])).shape
TensorShape([16, 32])
```

But for some use-cases, I'd like to explicitly make the Variable have a dynamic (`None`) shape:

```python
tf.Variable(tf.zeros([16, 32]), shape=[None, 32])
```

In the current beta, you can kind of do this via;
```python
import tensorflow as tf

class HasDynamicVariable(tf.Module):

  def __init__(self):
    super(HasDynamicVariable, self).__init__()
    self.v = None

  @tf.function(input_signature=[tf.TensorSpec([None, 2], tf.float32)])
  def make_dynamic_variable(self, initial_value):
    if self.v is None:
      self.v = tf.Variable(initial_value)
    return self.v
```
(xref https://github.com/tensorflow/community/pull/34#issuecomment-467690616)

But allowing you to explicitly annotate `tf.Variable(..., shape=...)` is IMO a lot cleaner and easier to use than having to declare it via the input signature.

(My particular use-case is to have something like a `tf.keras.layers.LSTM` with a dynamic batch size -- that means I'd like to do something like:

```
lstm = ...
lstm.reset_state(batch_size=5)
lstm(batch1)
lstm.reset_state(batch_size=20)
lstm(batch2)
```
)

**Comment by wangpengmit** [comment]
Linking similar issue #26104

**Comment by wangpengmit** [comment]
Fixed by https://github.com/tensorflow/tensorflow/commit/f9f2547b085969c67dc140aaa230551b6471a956

**Comment by liujian19911023** [comment]
does release less than 2.0 has this feature？

**Comment by wangpengmit** [comment]
It's available since 1.14
